### ISSUE TITLE
Dictionaries are compared by reference with `==`, unlike Arrays which are compared by value

### ISSUE THREAD

**Issue by AlexDarigan** [godotengine/godot#27615]
Godot 3.1

In Python (for example), if you were to compare Dictionaries they would evaluate to true as long as key-value pairs were the same. In GDScript, Dictionary comparison is based on a reference to the same Dictionary Object. 

```
var x = {0: 10}
var y = x
print(x == y) // Outputs true
```

```
var x = {0: 10}
var y = {0: 10}
print(x == y) // Outputs false
```

Is this the intended behaviour?

**Comment by Jummit** [comment]
This is intended. You look if the references of two dictionaries are the same, so creating two with the same values won't work.

**Comment by CodeDarigan** [comment]
> 
> 
> This is intended. You look if the references of two dictionaries are the same, so creating two with the same values won't work.

I mean, why do dicts comparison treat it via ref to the same object whereas Arrays are based on values.

**Comment by Zylann** [comment]
@CodeDarigan arrays used to were treated by ref too in the past, like many other languages do. Then there was requests to make it by value and now they are by value  ¯\\\_(ツ)_/¯

**Comment by Calinou** [comment]
This can be worked around with the following:

```gdscript
var x = {0: 10}
var y = {0: 10}
print(x.hash() == y.hash()) # Outputs `true`
```

**Comment by akien-mga** [comment]
See also #29222.

**Comment by razcore-art** [comment]
> This can be worked around thanks to the `Dictionary.hash()` method:
> 
> ```gdscript
> var x = {0: 10}
> var y = {0: 10}
> print(x.hash() == y.hash()) # Outputs `true`
> ```

Actually no, this doesn't quite work, cause the order matters since some other ppl complained that there wasn't order preservation with dicts.

So we get something like:

```gdscript
var d1 := {x = 0, y = 1}
var d2 := {y = 1, x = 0}
print(d1.hash() == d2.hash()) # False
```

Is there a way to preserver ordering and have the hashes be the same? I would think not, but honestly... I prefer unordered dicts because of this comparison issue.

Unordered dicts are the default in many languages.

**Comment by Calinou** [comment]
@razcore-art Dictionaries are ordered since Godot 3.0, but there's no way to opt out of that behvior.

**Comment by FeralBytes** [comment]
I just wanted to chime in that this recently bit me, as the order matters and so for complicated Dictionaries comparison via hash() is useless if just one element is out of order.

**Comment by MaaaxiKing** [comment]
I think dictionarys should be compared by value. It's just really confusing if you didn't hear before that they are compared by ref.

**Comment by Zylann** [comment]
Despite the fact I disagreed with all this by-value stuff that happened to arrays and dictionaries, and the fact dictionaries (famously being an unordered data structure) have been given ordered behavior which now causes unexpected problems with `hash()`, I think we should still allow comparing by reference, somehow, if we plan to replace the behavior of `==` (which btw will make `==` more expensive). 

Comparing by value can be done with some code, comparing by ref *cannot*. So some people are currently stuck because they need reference comparison: https://github.com/godotengine/godot/issues/33627
Besides, dictionaries and arrays behave by reference. We should still have tools to deal with that reality. Otherwise, how do you check equality if there is a cycle? You can't, but you could check by reference to detect it... which you can't do if comparison can only be done by value.
Issues like that happen already with arrays (it makes Godot lock up), and it will happen too if applied to dictionaries. 

Now, not saying we shouldn't do this (Python does it, so... it's possible). Just wanted to warn about the issues of such a decision.

**Comment by setzer22** [comment]
I came across this looking for a way to compare two dictionaries, so allow me to give my two cents.

Regardless of the behaviour of `==`, I'd say godot needs a standard function to compare dictionaries by value. In a dynamic language such as GDscript, many people use dictionaries as one-off objects when you don't want to create a class. Moreover: Dictionaries are more performant and memory efficient than classes/objects when you have many small objects. Dictionary comparison is in practice object comparison, which has many uses. Without a generic dictionary comparison operation, everyone will be forced to implement ad-hoc versions which may contain bugs.

GDScript could take the Java approach, and have `.equals()` for value comparisons and `==` for reference comparisons. Also, consistency, all Variant types should have this same `.equals()` API, which would be recursive for arrays and dictionaries. 

Also, IMO, having an inconsistent `==` for Arrays and Dictionaries is also a bad idea. Once of the most important principles in programming language design is consistency. It may be too late now to break the API, but if there's any chance to make this bevahiour consistent in a future GDScript version I'd seriously consider it.

Sorry if any of this came up as anything other than constructive criticism! GDScript is a very nice language and I only want to see it become better :)

**Comment by strank** [comment]
Just wanted to note something that hasn't been mentioned: most languages distinguish between identity comparison and equality comparison explicitly, plus in OO languages you also want a type comparison.
So to use python as an example:

- identity: `is` and `is not` (plus the `id()` function to make it explicit)
- equality: `==` and `!=` (plus ways for operator overloading with `__equals__` et al)
- type/subtype: `isinstance()` and `type()`

Gdscript is less clear:

- identity: `==` and `!=` but only for a few types when compared by reference, no `id()` function
- equality: `==` and `!=` for most things, as it simply passes it on to C++'s `==` which is an equality operator (C++ compares pointers for identity)
- type/subtype: `is`

I do like python's setup, as using `is` for identity feels very natural to me. Also, type checks are pretty rare in my experience, especially with gdscript's casting `as` returning `null` when it fails. And it seems good to make type checks explicit.

So I would vote for switching to the python way, but in the short term IMHO, the least disruptive way seems to be to add an `id` function to allow `id(a) == id(b)` for explicit identity checks, and change `==` to always compare equality.

**Comment by akien-mga** [comment]
Fixed by #35816.
